agents/ppo.py

```diff
@@ -88,88 +102,119 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
             'predicted_current_state_v_value': current_state_value_function
-        }
-
-    def learn(self):
-        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
-                self._node_idx,
-                self._trajectory_sample_size
-        ):
-            loss = self._trajectory_loss(trajectory)
-            # TODO backward here? or return to env?
-
-    def _trajectory_loss(self, trajectory):
-        reward = [data['reward'] for data in trajectory]
-        start_v_func, policy, node_idx, neighbor, destination, storage = trajectory[0]['extra_infos']
-        end_v_func = trajectory[-1]['extra_infos'][0]
-        return self._loss(node_idx, neighbor, destination, policy, start_v_func, reward, end_v_func)
+        })
+        # print(current_node_idx, destination_node_idx, current_state_value_function.detach())
+        # print(neighbor_node_ids.padded_values, neighbors_logits.detach())
+        return inputs
+
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
+            self._trajectory_sample_size,
+            self._trajectory_length
+        )
+        if not learn_trajectories:
+            return None
+        learn_norm_rewards = get_norm_rewards(learn_trajectories)
+        for trajectory, norm_rewards in zip(learn_trajectories, learn_norm_rewards):
+            loss += self._trajectory_loss(trajectory, norm_rewards)
+        loss /= self._trajectory_sample_size
+        self._optimizer.step(loss)
+        return loss.detach().item()
+
+    def _trajectory_loss(self, trajectory, norm_rewards):
+        reward = np.array(norm_rewards)
+
+        v_old, node_idx, neighbors, next_neighbor, neighbor_logits_old, destination, _ = trajectory[0]['extra_info']
+        _, _, _, _, _, _, end_v_old = trajectory[-1]['extra_info']
+        if trajectory[-1]['terminal']:
+            end_v_old = 0
+        _, neighbor_logits = self._actor(
+            current_node_idx=node_idx,
+            neighbor_node_ids=neighbors,
+            destination_node_idx=destination
+        )
+
+        logprob_old = self._get_logprob(neighbors, neighbor_logits_old)
+        logprob = self._get_logprob(neighbors, neighbor_logits)
+        entropy = Categorical(probs=torch.exp(logprob)).entropy()
+        next_logprob_old = logprob_old[neighbors.padded_values == torch.unsqueeze(next_neighbor, dim=1)]
+        next_logprob = logprob[neighbors.padded_values == torch.unsqueeze(next_neighbor, dim=1)]
+
+        v = self._critic(
+            current_node_idx=node_idx,
+            destination_node_idx=destination
+        )
+
+        return self._loss(next_logprob, next_logprob_old, v, v_old, end_v_old, reward, entropy)
+
+    def _get_logprob(self, neighbors, neighbors_logits):
+        inf_tensor = torch.zeros(neighbors_logits.shape)
+        inf_tensor[~neighbors.mask] = -1e10
```

Comment:
Put `-1e10` in the constant or use `-torch.inf` (Please check the last option, I am not sure it will work as intended)

Reply:
I tried to use -inf and something strange hapend, let's make constant